### PR TITLE
Adding avoid_look_at task map

### PR DESCRIPTION
**Pull request by cmower**
* `avoid_look_at` avoids pointing the end-effector at a given spherical object
* adding `.cpp`/`.h`/`.in` files, compiles and runs as expected
* updated `CMakeLists.txt` and `exotica_plugins.xml`
* added unit test
* added picture for doxygen
* code formatting

# Checklist

- [x] code formatting: run `find -name '*.cpp' -o -name '*.h' -o -name '*.hpp' | xargs clang-format-3.9 -style=file -i` in the root directory

### CODE REVIEW

**Comment by cmower** [comment]
Updated tests and ensured they build locally. Assuming CI goes green I will squash and then should all be good to merge.

File: exotations/exotica_core_task_maps/test/test_maps.cpp [hunk 1]
```diff
@@ -1159,11 +1159,37 @@ TEST(ExoticaTaskMaps, testManipulability)
...
+{
+    try
+    {
+        TEST_COUT << "AvoidLookAt Test";
+        Initializer map("exotica/AvoidLookAt", {{"Name", std::string("MyTask")},
+						{"EndEffector", std::vector<Initializer>(
+											 {Initializer("EffPoint", {{"Link", std::string("EffPoint")},
+														   {"Base", std::string("base")},
+														   {"BaseOffset", std::string("0 0 1")}}),
+											  Initializer("EffPoint", {{"Link", std::string("Obj")},
```
Comment (VladimirIvan) [inline exotations/exotica_core_task_maps/test/test_maps.cpp]: Should this `Initializer("EffPoint",` be this `Initializer("Frame",` ?

Comment (cmower) [reply exotations/exotica_core_task_maps/test/test_maps.cpp]: Ahh yes, will update

File: exotations/exotica_core_task_maps/CMakeLists.txt [hunk 2]
```diff
@@ -90,4 +92,4 @@ if(CATKIN_ENABLE_TESTING)
   catkin_add_gtest(test_maps test/test_maps.cpp)
   target_link_libraries(test_maps ${catkin_LIBRARIES})
   add_dependencies(test_maps ${catkin_EXPORTED_TARGETS})
-endif()
\ No newline at end of file
+endif()
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/CMakeLists.txt]: Please do not change file endings. We should have an empty line at end of file.

File: exotations/exotica_core_task_maps/exotica_plugins.xml [hunk 3]
```diff
@@ -62,4 +62,7 @@
   <class name="exotica/Manipulability" type="exotica::Manipulability" base_class_type="exotica::TaskMap">
     <description>Maximizes the manipulability measure using Yoshikawa's measure based on the velocity ellipsoid. Note, in-order to maximize manipulability the associated Rho must be negative. Derivatives are computed using finite differencing.</description>
   </class>
-</library>
\ No newline at end of file
+  <class name="exotica/AvoidLookAt" type="exotica::AvoidLookAt" base_class_type="exotica::TaskMap">
+    <description>Avoids looking at given target points. This can only be used as an inequality constraint.</description>
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/exotica_plugins.xml]: Is there a way to make this a cost, e.g., using an epsilon?

File: exotations/exotica_core_task_maps/include/exotica_core_task_maps/avoid_look_at.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    virtual ~AvoidLookAt();
+
+    void Instantiate(AvoidLookAtInitializer& init) override;
+    void Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi) override;
+    void Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef jacobian) override;
+    int TaskSpaceDim() override;
+
+private:
+  int n_objects_;  ///< Number of spherical objects.
+  Eigen::VectorXd radii2_; ///< The square of each object radii.
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/include/exotica_core_task_maps/avoid_look_at.h]: Rename to `radii_squared_` or `squared_radii_`. `radii2_` seems more like `the second copy of that variable`.

File: exotations/exotica_core_task_maps/src/avoid_look_at.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+REGISTER_TASKMAP_TYPE("AvoidLookAt", exotica::AvoidLookAt);
+
+namespace exotica
+{
+AvoidLookAt::AvoidLookAt() = default;
+AvoidLookAt::~AvoidLookAt() = default;
+
+void AvoidLookAt::Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi)
+{
+
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/src/avoid_look_at.cpp]: Empty lines need to go. Also formatting isn't applied.

File: exotations/exotica_core_task_maps/src/avoid_look_at.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+namespace exotica
+{
+AvoidLookAt::AvoidLookAt() = default;
+AvoidLookAt::~AvoidLookAt() = default;
+
+void AvoidLookAt::Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi)
+{
+  
+
+  Eigen::Vector3d c = Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(0).p.data);
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/src/avoid_look_at.cpp]: `const Eigen::Vector3d`

File: exotations/exotica_core_task_maps/src/avoid_look_at.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  for (int i = 0; i < n_objects_; ++i) {
+    Eigen::Vector3d o = Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(i+1).p.data);
+    phi(i) = 0.5 * (radii2_(i) - (c * o.dot(c) / c.squaredNorm() - o).squaredNorm());
+  }
+   
+}
+
+void AvoidLookAt::Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef jacobian)
+{
+
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/src/avoid_look_at.cpp]: Empty line needs to go.

File: exotations/exotica_core_task_maps/src/avoid_look_at.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  Eigen::Vector3d c = Eigen::Map<Eigen::Vector3d>(kinematics[0].Phi(0).p.data);
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/src/avoid_look_at.cpp]: `const`

File: exotations/exotica_core_task_maps/src/avoid_look_at.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      jacobian(i, j) = (od - c * od.dot(c)/c_squared_norm).dot(a - o);
+    }
+    
+  }
+  
+}
+
+void AvoidLookAt::Instantiate(AvoidLookAtInitializer& init)
+{
+
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/src/avoid_look_at.cpp]: Empty line

File: exotations/exotica_core_task_maps/src/avoid_look_at.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  if (frames_.size() < 2) ThrowNamed("A constant point defined in the end-effector frame is required.");
+  n_objects_ = init.Radii.size();
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/src/avoid_look_at.cpp]: Check that frames and Radii match/have a correlation, e.g., `Radii.size() == frames_.size() - 1`

File: exotations/exotica_core_task_maps/test/test_maps.cpp [hunk 11]
```diff
@@ -1159,11 +1159,37 @@ TEST(ExoticaTaskMaps, testManipulability)
...
+											 {Initializer("Frame", {{"Link", std::string("EffPoint")},
+														   {"Base", std::string("base")},
+														   {"BaseOffset", std::string("0 0 1")}}),
+											  Initializer("Frame", {{"Link", std::string("Obj")},
+														   {"Base", std::string("base")},
+														   {"BaseOffset", std::string("0 0 1")}})})}});
+
+        UnconstrainedEndPoseProblemPtr problem = setup_problem(map);
+        EXPECT_TRUE(test_random(problem));
+        // TODO: Add test_values
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/test/test_maps.cpp]: I'd recommend adding values for this task map.

File: exotations/exotica_core_task_maps/test/test_maps.cpp [hunk 12]
```diff
@@ -1159,11 +1159,37 @@ TEST(ExoticaTaskMaps, testManipulability)
...
 int main(int argc, char** argv)
 {
     testing::InitGoogleTest(&argc, argv);
     ros::init(argc, argv, "EXOTica_test_maps");
     int ret = RUN_ALL_TESTS();
     Setup::Destroy();
     return ret;
-}
\ No newline at end of file
+}
```
Comment (wxmerkt) [inline exotations/exotica_core_task_maps/test/test_maps.cpp]: End of file change.

**Comment by wxmerkt** [comment]
@cmower What is the status of this pull request?

**Comment by cmower** [comment]
This is not a priority for me at the moment. I will come back and finish some point in next few weeks.

File: exotations/exotica_core_task_maps/exotica_plugins.xml (hunk 3 above)
Comment (cmower) [reply exotations/exotica_core_task_maps/exotica_plugins.xml]: Using as cost is possible by modifying as potential field. Will add this in next commit.

**Comment by wxmerkt** [comment]
Okay, thank you for getting back - I will close the PR then. Please reopen once you've addressed the feedback above.

**Pull request closed without merging**